datalad/plugin/addurls.py

```diff
@@ -0,0 +1,654 @@
+                                     unit=" Files")
+    pbar_addurl.start()
+
+    files_to_add = []
+    meta_to_add = []
+    addurl_results = []
+    for row_idx, row in enumerate(rows, 1):
+        pbar_addurl.update(row_idx)
+        fname_abs = os.path.join(dataset.path, row["filename"])
+        if row["subpath"]:
+            # Adjust the dataset and filename for an `addurl` call
+            # from within the subdataset that will actually contain
+            # the link.
+            ds_current = Dataset(os.path.join(dataset.path, row["subpath"]))
+            ds_filename = os.path.relpath(fname_abs, ds_current.path)
+        else:
+            ds_current = dataset
+            ds_filename = row["filename"]
+
+        if os.path.exists(fname_abs) or os.path.islink(fname_abs):
+            if ifexists == "skip":
+                addurl_results.append(
+                    get_status_dict(action="addurls",
+                                    ds=ds_current,
+                                    type="file",
+                                    path=fname_abs,
+                                    status="notneeded"))
+                continue
+            elif ifexists == "overwrite":
+                lgr.debug("Removing %s", fname_abs)
+                os.unlink(fname_abs)
+            else:
+                lgr.debug("File %s already exists", fname_abs)
+
+        try:
+            ds_current.repo.add_url_to_file(ds_filename, row["url"],
+                                            batch=True, options=annex_options)
+        except AnnexBatchCommandError:
+            addurl_results.append(
+                get_status_dict(action="addurls",
+                                ds=ds_current,
+                                type="file",
+                                path=os.path.join(ds_current.path,
+                                                  ds_filename),
+                                message="failed to add URL",
+                                status="error"))
+
+            continue
+        # Collect the status dicts to yield later so that we don't
+        # interrupt the progress bar.
+        addurl_results.append(
+            get_status_dict(action="addurls",
+                            ds=ds_current,
+                            type="file",
+                            path=os.path.join(ds_current.path,
+                                              ds_filename),
+                            status="ok"))
+
+        files_to_add.append(row["filename"])
+        meta_to_add.append((ds_current, ds_filename, row["meta_args"]))
+    pbar_addurl.finish()
+
+    for result in addurl_results:
+        yield result
+
+        msg = message or """\
+[DATALAD] add files from URLs
+
+url_file='{}'
+url_format='{}'
+filename_format='{}'""".format(url_file, url_format, filename_format)
+
+    if files_to_add:
+        for r in dataset.add(files_to_add, message=msg):
```

Comment:
recent "leaning" of imports hurts us here:
```
$> datalad plugin addurls url_file=phenotype_mri_head10.csv url_format='{MRIs}' filename_format='{project}/sub-{participant_id}/ses-{session_id}_rep-{_repindex}/file' fast=True                                                                  
addurls(ok): /tmp/testaddurl/abide_initiative/sub-50002/ses-1_rep-0/file (file)                                          
addurls(ok): /tmp/testaddurl/abide_initiative/sub-50003/ses-1_rep-0/file (file)
addurls(ok): /tmp/testaddurl/abide_initiative/sub-50004/ses-1_rep-0/file (file)
...
[ERROR  ] 'Dataset' object has no attribute 'add' [addurls.py:dlplugin:625] (AttributeError) 
```
do
```python
from datalad.distribution.add import Add
```
to get `.add` bound

Reply:
Thanks.  I should have ran it from the command line after rebasing rather than taking the passing tests as a green light :/